command/agent/config.go

```diff
@@ -747,61 +793,28 @@ func normalizeBind(addr, bind string) string {
 // is resolved and returned with the port.
 //
 // Loopback is only considered a valid advertise address in dev mode.
-func normalizeAdvertise(addr string, bind string, defport int, dev bool) (string, error) {
+func normalizeAdvertise(addr string, bind string, defport int) (string, error) {
 	if addr != "" {
 		// Default to using manually configured address
-		_, _, err := net.SplitHostPort(addr)
+		host, port, err := net.SplitHostPort(addr)
 		if err != nil {
 			if !isMissingPort(err) {
 				return "", fmt.Errorf("Error parsing advertise address %q: %v", addr, err)
 			}
-
-			// missing port, append the default
-			return net.JoinHostPort(addr, strconv.Itoa(defport)), nil
+			host = addr
+			port = strconv.Itoa(defport)
 		}
-		return addr, nil
-	}
 
-	// Fallback to bind address first, and then try resolving the local hostname
-	ips, err := net.LookupIP(bind)
-	if err != nil {
-		return "", fmt.Errorf("Error resolving bind address %q: %v", bind, err)
-	}
-
-	// Return the first unicast address
-	for _, ip := range ips {
-		if ip.IsLinkLocalUnicast() || ip.IsGlobalUnicast() {
-			return net.JoinHostPort(ip.String(), strconv.Itoa(defport)), nil
-		}
-		if ip.IsLoopback() && dev {
-			// loopback is fine for dev mode
-			return net.JoinHostPort(ip.String(), strconv.Itoa(defport)), nil
+		ipStr, err := parseSingleIPTemplate(host)
+		if err != nil {
+			return "", fmt.Errorf("Error parsing advertise address template: %v", err)
 		}
-	}
-
-	// As a last resort resolve the hostname and use it if it's not
-	// localhost (as localhost is never a sensible default)
-	host, err := os.Hostname()
-	if err != nil {
-		return "", fmt.Errorf("Unable to get hostname to set advertise address: %v", err)
-	}
 
-	ips, err = net.LookupIP(host)
-	if err != nil {
-		return "", fmt.Errorf("Error resolving hostname %q for advertise address: %v", host, err)
+		return net.JoinHostPort(ipStr, port), nil
 	}
 
-	// Return the first unicast address
-	for _, ip := range ips {
-		if ip.IsLinkLocalUnicast() || ip.IsGlobalUnicast() {
-			return net.JoinHostPort(ip.String(), strconv.Itoa(defport)), nil
-		}
-		if ip.IsLoopback() && dev {
-			// loopback is fine for dev mode
-			return net.JoinHostPort(ip.String(), strconv.Itoa(defport)), nil
-		}
-	}
-	return "", fmt.Errorf("No valid advertise addresses, please set `advertise` manually")
+	// Fallback to bind address, as it has been resolved before.
+	return net.JoinHostPort(bind, strconv.Itoa(defport)), nil
```

Comment:
FWIW, Consul will be likely making a change to its defaults in time for its 0.8 release to use `GetPrivateIP` as its default bind address.  And yes, listening on multiple interfaces is something we all want to do (though I'd argue it should be a UNIX socket in /tmp vs `127.0.0.1` that way filesystem permissions could be applied if necessary, and you can figure out the UID of who is connecting to you via a unix socket, but not via loopback - but more on that when we get closer to multiple listeners).